packages/database/src/ModelRegistry.js

```diff
@@ -83,8 +84,20 @@ export class ModelRegistry {
   }
 
   async wrapInTransaction(wrappedFunction) {
-    return this.database.wrapInTransaction(transactingDatabase => {
-      const transactingModelRegistry = new ModelRegistry(transactingDatabase, this.modelClasses);
+    return this.database.wrapInTransaction(async transactingDatabase => {
+      const schemata = {};
+      await Promise.all(
+        Object.keys(this.modelClasses).map(async modelName => {
+          const modelKey = getModelKey(modelName);
+          schemata[modelKey] = await this[modelKey].fetchSchema();
+        }),
+      );
```

Comment:
Ah so by pre-fetching the schemas, we save having those queries running in the background whenever we start a transaction.

I just wanted to check about the generality of this issue. Will this kind of issue pop up any time we have an existing query running in a transaction, and then throw an error before that query resolves?
eg.
```
db.wrapInTransaction(transactingDb => {
  transactingDb.runSql("// Some long running query");
  throw new Error('Error out before query finishes');
});
```

Reply:
@edmofro do you wanna chime in on this one? Probably a bit more in your wheelhouse!

Reply:
@rohan-bes yes, if that query is not awaited
There was a subtle issue here where:
- We start the schema fetch query in the constructor, which can't await
- We then await the schema fetch query any time we come to actually use a model, to get around the above
- When we start this transaction, we do some work before the first time we use a model, and that work can throw an error and bail out of the transaction - then when the unawaited query completes, its transaction has already been rolled back
  - (this is why @alexd-bes's original fix of moving the sometimes-crashing line of code to below the first fetch worked)
So long as we're awaiting queries, as we almost invariably do, this shouldn't happen other places.

Reply:
Thanks for the full explanation @edmofro :pray: 

Thoughts on introducing some class to maintain a list of all running transaction queries, and we ensure that all have been awaited before throwing an error?

Something like:
```
  // In TupaiaDatabase#wrapInTransaction
  wrapInTransaction(wrappedFunction) {
    const transactionQueryList = new TransactionQueryList(); 
    return this.connection.transaction(async (transaction) => {
      let errorInTransaction;
      try {
        await wrappedFunction(new TupaiaDatabase(transaction, this.changeChannel, transactionQueryList));
      } catch (e) {
        errorInTransaction = e;
      } finally {
        await transactionQueryList.allQueriesCompleted();
      }
      
      if (errorInTransaction) {
        throw errorInTransaction;
      }
    });
  }
```
Then it's a matter of finding the places were make queries in `TupaiaDatabase` and adding those queries into the `transactionQueryList`.

I've definitely seen this issue crop up at least once before. It could have been caused by the same issue with the schemas though, I never worked out the cause. But I wouldn't mind having a general fix for this.

Reply:
But I don't feel too strongly about it. Happy to approve this :+1:

Reply:
Yeah honestly I feel that's a little heavy on the safety. When this does happen it does something like _crash the whole server_ and then we figure out what we're failing to await. I personally prefer fixing any time it crops up - in this case it was definitely an overall improvement in other ways too!